rust/rti/src/trace.rs

```diff
@@ -0,0 +1,631 @@
+
+ * @section DESCRIPTION
+ * Definitions of tracepoint events for use with the C code generator and any other
+ * code generator that uses the C infrastructure (such as the Python code generator).
+ *
+ * See: https://www.lf-lang.org/docs/handbook/tracing?target=c
+ *
+ * The trace file is named trace.lft and is a binary file with the following format:
+ *
+ * Header:
+ * * instant_t: The start time. This is both the starting physical time and the starting logical time.
+ * * int: Size N of the table mapping pointers to descriptions.
+ * This is followed by N records each of which has:
+ * * A pointer value (the key).
+ * * A null-terminated string (the description).
+ *
+ * Traces:
+ * A sequence of traces, each of which begins with an int giving the length of the trace
+ * followed by binary representations of the trace_record struct written using fwrite().
+ */
+use std::fs::{File, OpenOptions};
+use std::io::Write;
+use std::sync::{Arc, RwLock};
+
+use crate::tag::{Instant, Interval, Microstep, Tag};
+use crate::RTIRemote;
+
+use zerocopy::AsBytes;
+
+type Void = i64;
+
+const TRACE_BUFFER_CAPACITY: u32 = 2048;
+
+#[derive(Clone, Copy, Debug)]
+pub enum TraceEvent {
+    ReceiveAdrAd,
+    ReceiveAdrQr,
+    ReceiveFedId,
+    ReceiveNet,
+    ReceiveLtc,
+    ReceivePortAbs,
+    ReceiveResign,
+    ReceiveStopReq,
+    ReceiveStopReqRep,
+    ReceiveTaggedMsg,
+    ReceiveTimestamp,
+    ReceiveUnidentified,
+    SendAck,
+    SendPTag,
+    SendReject,
+    SendStopGrn,
+    SendStopReq,
+    SendTag,
+    SendTaggedMsg,
+    SendTimestamp,
+}
+
+impl TraceEvent {
+    pub fn to_value(&self) -> u64 {
+        match self {
+            TraceEvent::ReceiveAdrAd => 51,
+            TraceEvent::ReceiveAdrQr => 52,
+            TraceEvent::ReceiveFedId => 40,
+            TraceEvent::ReceiveNet => 35,
+            TraceEvent::ReceiveLtc => 36,
+            TraceEvent::ReceivePortAbs => 45,
+            TraceEvent::ReceiveResign => 44,
+            TraceEvent::ReceiveStopReq => 37,
+            TraceEvent::ReceiveStopReqRep => 38,
+            TraceEvent::ReceiveTaggedMsg => 47,
+            TraceEvent::ReceiveTimestamp => 34,
+            TraceEvent::ReceiveUnidentified => 53,
+            TraceEvent::SendAck => 11,
+            TraceEvent::SendPTag => 20,
+            TraceEvent::SendReject => 22,
+            TraceEvent::SendStopGrn => 18,
+            TraceEvent::SendStopReq => 16,
+            TraceEvent::SendTag => 21,
+            TraceEvent::SendTaggedMsg => 26,
+            TraceEvent::SendTimestamp => 13,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct TraceRecord {
+    event_type: u64,
+    pointer: Void, // pointer identifying the record, e.g. to self struct for a reactor.
+    src_id: i32,   // The ID number of the source (e.g. worker or federate) or -1 for no ID number.
+    dst_id: i32, // The ID number of the destination (e.g. reaction or federate) or -1 for no ID number.
```

Comment:
It's very minor, but the indentation of the comment at this line seems off.

Reply:
Thank you! I'll apply `cargo fmt` again.